### ISSUE TITLE
SDPA produces NaN with padding mask

### ISSUE THREAD

**Issue by carmocca** [pytorch/pytorch#103749]
### 🐛 Describe the bug

I'm implementing padding support directly on my LLM model. To do so, I add extra rows to the boolean attention mask with all `False` values.

However, calling `torch.nn.functional.scaled_dot_product_attention` returns NaNs.

An equivalent and naive implementation also does (see `naive_fsdpa_1`).
But if I use `torch.finfo(att.dtype).min` as the masked fill value, the result is correct (see `naive_fsdpa_2`)

Here's the repro:

```python
import torch

scale = 0.125

# 5 tokens, 1 padding
seq_length = 5
max_seq_length = 10

q = torch.randn(1, 2, seq_length + 1, 8)
k = torch.randn(1, 2, max_seq_length, 8)
v = torch.randn(1, 2, max_seq_length, 8)
mask = torch.tensor([[[
    [ True, False, False, False, False, False, False, False, False, False],  # regular mask
    [ True,  True, False, False, False, False, False, False, False, False],
    [ True,  True,  True, False, False, False, False, False, False, False],
    [ True,  True,  True,  True, False, False, False, False, False, False],
    [ True,  True,  True,  True,  True, False, False, False, False, False],
    [False, False, False, False, False, False, False, False, False, False]]]])  # padding mask



def torch_sdpa(q, k, v, mask):
    return torch.nn.functional.scaled_dot_product_attention(q, k, v, mask, dropout_p=0.0, scale=scale)


def naive_sdpa_1(q, k, v, mask):
    att = (q @ k.transpose(-2, -1)) * scale
    att = torch.masked_fill(att, ~mask, float("-inf"))
    att = torch.nn.functional.softmax(att, dim=-1)
    return att @ v


def naive_sdpa_2(q, k, v, mask):
    att = (q @ k.transpose(-2, -1)) * scale
    att = torch.masked_fill(att, ~mask, torch.finfo(att.dtype).min)
    att = torch.nn.functional.softmax(att, dim=-1)
    return att @ v

y = torch_sdpa(q, k, v, mask)
print(torch.isnan(y).any())

y = naive_sdpa_1(q, k, v, mask)
print(torch.isnan(y).any())

y = naive_sdpa_2(q, k, v, mask)
print(torch.isnan(y).any())
```

This seems to suggest that there's a precision issue. Is this expected or known? Could `scaled_dot_product_attention` be fixed?

Side question: do I lose flash attn if I append the padding mask like this?

Thank you for your help.

Maybe linked: https://github.com/pytorch/pytorch/issues/101967

### Versions

Current master

cc @ezyang @gchanan @zou3519 @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @msaroufim @wconstab @bdhirsh @anijain2305

**Comment by professorabhay** [comment]
`import torch

scale = 0.125

# 5 tokens, 1 padding
seq_length = 5
max_seq_length = 10

q = torch.randn(1, 2, seq_length + 1, 8)
k = torch.randn(1, 2, max_seq_length, 8)
v = torch.randn(1, 2, max_seq_length, 8)
mask = torch.tensor([[[
    [True, False, False, False, False, False, False, False, False, False],  # regular mask
    [True, True, False, False, False, False, False, False, False, False],
    [True, True, True, False, False, False, False, False, False, False],
    [True, True, True, True, False, False, False, False, False, False],
    [True, True, True, True, True, False, False, False, False, False],
    [False, False, False, False, False, False, False, False, False, False]]]])  # padding mask


def torch_sdpa(q, k, v, mask):
    return torch.nn.functional.scaled_dot_product_attention(q, k, v, mask, dropout_p=0.0, scale=scale)


def naive_sdpa_1(q, k, v, mask):
    att = (q @ k.transpose(-2, -1)) * scale
    att = torch.where(~mask, float("-inf"), att)
    att = torch.nn.functional.softmax(att, dim=-1)
    return att @ v


def naive_sdpa_2(q, k, v, mask):
    att = (q @ k.transpose(-2, -1)) * scale
    att = torch.where(~mask, torch.finfo(att.dtype).min, att)
    att = torch.nn.functional.softmax(att, dim=-1)
    return att @ v


y = torch_sdpa(q, k, v, mask)
print(torch.isnan(y).any())

y = naive_sdpa_1(q, k, v, mask)
print(torch.isnan(y).any())

y = naive_sdpa_2(q, k, v, mask)
print(torch.isnan(y).any())
`
I would suggest you to try this code once and let me know @carmocca

**Comment by professorabhay** [comment]
`import torch

scale = 0.125

# 5 tokens, 1 padding
seq_length = 5
max_seq_length = 10

q = torch.randn(1, 2, seq_length + 1, 8)
k = torch.randn(1, 2, max_seq_length, 8)
v = torch.randn(1, 2, max_seq_length, 8)
mask = torch.tensor([[[
    [True, False, False, False, False, False, False, False, False, False],  # regular mask
    [True, True, False, False, False, False, False, False, False, False],
    [True, True, True, False, False, False, False, False, False, False],
    [True, True, True, True, False, False, False, False, False, False],
    [True, True, True, True, True, False, False, False, False, False],
    [False, False, False, False, False, False, False, False, False, False]]]])  # padding mask


def torch_sdpa(q, k, v, mask):
    return torch.nn.functional.scaled_dot_product_attention(q, k, v, mask, dropout_p=0.0, scale=scale)


def naive_sdpa_1(q, k, v, mask):
    att = (q @ k.transpose(-2, -1)) * scale
    att = torch.where(~mask, float("-inf"), att)
    att = torch.nn.functional.softmax(att, dim=-1)
    return att @ v


def naive_sdpa_2(q, k, v, mask):
    att = (q @ k.transpose(-2, -1)) * scale
    att = torch.where(~mask, torch.finfo(att.dtype).min, att)
    att = torch.nn.functional.softmax(att, dim=-1)
    return att @ v


y = torch_sdpa(q, k, v, mask)
print(torch.isnan(y).any())

y = naive_sdpa_1(q, k, v, mask)
print(torch.isnan(y).any())

y = naive_sdpa_2(q, k, v, mask)
print(torch.isnan(y).any())
`


I would suggest you to try this code once and let me know @carmocca

**Comment by professorabhay** [comment]
I would suggest you to try this code once and let me know @carmocca -> 

`import torch`

`scale = 0.125`

`# 5 tokens, 1 padding
seq_length = 5
max_seq_length = 10`

`q = torch.randn(1, 2, seq_length + 1, 8)
k = torch.randn(1, 2, max_seq_length, 8)
v = torch.randn(1, 2, max_seq_length, 8)
mask = torch.tensor([[[
    [True, False, False, False, False, False, False, False, False, False],  # regular mask
    [True, True, False, False, False, False, False, False, False, False],
    [True, True, True, False, False, False, False, False, False, False],
    [True, True, True, True, False, False, False, False, False, False],
    [True, True, True, True, True, False, False, False, False, False],
    [False, False, False, False, False, False, False, False, False, False]]]])  # padding mask`


`def torch_sdpa(q, k, v, mask):
    return torch.nn.functional.scaled_dot_product_attention(q, k, v, mask, dropout_p=0.0, scale=scale)`


`def naive_sdpa_1(q, k, v, mask):
    att = (q @ k.transpose(-2, -1)) * scale
    att = torch.where(~mask, float("-inf"), att)
    att = torch.nn.functional.softmax(att, dim=-1)
    return att @ v`


`def naive_sdpa_2(q, k, v, mask):
    att = (q @ k.transpose(-2, -1)) * scale
    att = torch.where(~mask, torch.finfo(att.dtype).min, att)
    att = torch.nn.functional.softmax(att, dim=-1)
    return att @ v`


`y = torch_sdpa(q, k, v, mask)
print(torch.isnan(y).any())`

`y = naive_sdpa_1(q, k, v, mask)
print(torch.isnan(y).any())`

`y = naive_sdpa_2(q, k, v, mask)
print(torch.isnan(y).any())`

**Comment by mingfeima** [comment]
@Valentine233 could you please help debug this issue ?

**Comment by Valentine233** [comment]
> @Valentine233 could you please help debug this issue ?

Sure.

**Comment by drisspg** [comment]
Initially responded to #103963 with:
Correct, the problem is a completely masked out row. This is indicating that this batch, head query entry should attend to nothing in the keys which doesn't make the most sense.
![image](https://github.com/pytorch/pytorch/assets/32754868/d336d267-c99d-4f4a-a4e4-c98320429e1f)
Which causes the indeterminate: 0/0 fo the softmax calc.

The motivation for the boolean mask was to help people with their padding, and masking code which is typically used for creating a batch of varying sequences to perform SDPA on. They otheroption, encouraged for inference, is to use NestedTensors. This function is still in Beta. And I would be curious for arguments against the boolean mask.

**Comment by drisspg** [comment]
@nikitaved Responded with: 
>"@drisspg , the point being is that boolean mask result makes sense, the floating point mask does not... To me, at least..."

**Comment by nikitaved** [comment]
For the context. Boolean mask behavior makes sense to me. What is the reason behind floating point masks behaving differently as per the following example?

```python
In [1]: import torch

In [2]: import torch.nn.functional as F

In [3]: q = torch.eye(32, 32, device='cuda').expand(1, 1, 32, 32).contiguous()

In [4]: k = torch.eye(32, 32, device='cuda').expand(1, 1, 32, 32).contiguous()

In [5]: v = torch.eye(32, 32, device='cuda').expand(1, 1, 32, 32).contiguous()

In [6]: attn_mask = torch.ones(1, 1, 32, 32, device='cuda').to(torch.bool)

In [7]: attn_mask[..., :16, 16:] = 0

In [8]: res1 = F.scaled_dot_product_attention(q, k, v, attn_mask)

In [9]: res2 = F.scaled_dot_product_attention(q, k, v, attn_mask.to(torch.float))

In [10]: res1
Out[10]: 
tensor([[[[0.0737, 0.0618, 0.0618,  ..., 0.0000, 0.0000, 0.0000],
          [0.0618, 0.0737, 0.0618,  ..., 0.0000, 0.0000, 0.0000],
          [0.0618, 0.0618, 0.0737,  ..., 0.0000, 0.0000, 0.0000],
          ...,
          [0.0311, 0.0311, 0.0311,  ..., 0.0371, 0.0311, 0.0311],
          [0.0311, 0.0311, 0.0311,  ..., 0.0311, 0.0371, 0.0311],
          [0.0311, 0.0311, 0.0311,  ..., 0.0311, 0.0311, 0.0371]]]],
       device='cuda:0')

In [11]: res2
Out[11]: 
tensor([[[[0.0540, 0.0453, 0.0453,  ..., 0.0167, 0.0167, 0.0167],
          [0.0453, 0.0540, 0.0453,  ..., 0.0167, 0.0167, 0.0167],
          [0.0453, 0.0453, 0.0540,  ..., 0.0167, 0.0167, 0.0167],
          ...,
          [0.0311, 0.0311, 0.0311,  ..., 0.0371, 0.0311, 0.0311],
          [0.0311, 0.0311, 0.0311,  ..., 0.0311, 0.0371, 0.0311],
          [0.0311, 0.0311, 0.0311,  ..., 0.0311, 0.0311, 0.0371]]]],
       device='cuda:0')
```

**Comment by drisspg** [comment]
My follow up would be, I the float version also makes sense to me. You are turning all the False values to 0 when you convert from boolean to float. So your attn _mask will look like the following.
<img width="436" alt="Screenshot 2023-06-21 at 10 22 41 AM" src="https://github.com/pytorch/pytorch/assets/32754868/3591b064-8108-47dd-b13b-3c70e907bc0d">

The "0" row is indicating that, from the softmax calc, all entries in the key should be attended too equally

**Comment by drisspg** [comment]
@nikita can you make this an easier snipppet to run?

**Comment by drisspg** [comment]
@nikitaved I would be happy to take a look at this example but can you make it a python script and not the output of a jupyter notebook?

**Comment by danieldk** [comment]
> For the context. Boolean mask behavior makes sense to me. What is the reason behind floating point masks behaving differently as per the following example?

@nikitaved  The floating point mask does work correctly. From the docs:

> A float mask of the same type as query, key, value that **is added** to the attention score.

In order to mask out the the values for softmax, you need to set them to a very large negative value. So, the correct mask is not

```
attn_mask.to(torch.float)
```

But something like:

```
(1.0 - self.attn_mask.float()) * torch.finfo(dtype).min
```

In fact, PyTorch does something similar internally, converting the bool mask to floats. Translated to Python code [it does the following](https://github.com/pytorch/pytorch/blob/223f23292873940d1028730fb381d6cf73aac404/aten/src/ATen/native/transformers/attention.cpp#L673) (it uses a masked fill, but same principle):

```
(1.0 - self.attn_mask.float()) * float("-inf")
```

However, this is where the issue arises. If a token has an attention mask consisting of only `False` (does not attend to another token), softmax will return `NaN`. However, if you use the minimum value for a dtype, it will return a uniform distribution.

**Comment by danieldk** [comment]
> This is indicating that this batch, head query entry should attend to nothing in the keys which doesn't make the most sense.

It does make sense once you start combining attention masks and causal masks. You cannot pass both to most of the Torch SDP implementation, plus causal masks are currently broken if the key and value lengths are not the same (#103082). So the only way to use Torch SDP currently is to combine the causal and attention masks and pass the intersecteon to `scaled_dot_product_attention`.

These all-false masks occur in that scenario. Suppose that a sequence has padding at the end. Then the last padding piece does not attend to earlier tokens (due to the causal mask). That by itself would be fine, because the piece would still attend to itself. However, since it is padding, it is also masked out by the attention mask. So, the result is a piece/query that does not attend to any key.

This is how we ran into this issue. However, now we are using the float mask and mask the logits by setting them to `torch.finfo(dtype).min` rather than `-inf`.

**Comment by nikitaved** [comment]
@danieldk, by problem is that `attn_mask.to(torch.bool)` and `attn_mask.to(torch.float)` being not semantically equivalent even if `attn_mask` is just `{0, 1}` binary. Besides, in the floating case, I struggle to understand why `attn_mask` is a mask, not a weight, since it is not used for masking...

**Comment by danieldk** [comment]
> @danieldk, by problem is with `attn_mask.to(torch.bool)` and `attn_mask.to(torch.float)` not being semantically equivalent even if `attn_mask` is just `{0, 1}` binary.

Most libraries mask logits by adding a large negative value, so it does make it easy to plug into existing libraries. If they were semantically the same, it wouldn't make sense to accept float tensors at all.

**Comment by nikitaved** [comment]
@danieldk , the issues is with `softmax`, so, it seems, it is safe to close this issue...

**Comment by nikitaved** [comment]
@cpuhrsch , the difference between boolean and floating point `attn_mask` means that:
- if `attn_mask` is boolean, we have all the methods to implement SDPA with BSR masks.
- if `attn_mask` is floating point, we do not need any of the functionality we implemented before (bsr_softmax, sampled_addmm), but we need `add(dense, bsr)`.

Unless boolean and floating point semantics could be unified...

**Comment by danieldk** [comment]
> What are these libraries may I ask, just curious?

It's pretty common. One example is Hugging Face Transformers:

https://github.com/huggingface/transformers/blob/6ce6d62b6f20040129ec9831e7c4f6576402ea42/src/transformers/modeling_utils.py#L902
https://github.com/huggingface/transformers/blob/6ce6d62b6f20040129ec9831e7c4f6576402ea42/src/transformers/models/bert/modeling_bert.py#L355

**Comment by nikitaved** [comment]
@danieldk , all right, that makes sense to me. Thank you for the examples!

**Comment by nikitaved** [comment]
@danieldk , there is an issue about `softmax` which is not closed, https://github.com/pytorch/pytorch/issues/25110.
I think we can still fix it if proposition https://github.com/pytorch/pytorch/issues/25110#issuecomment-1601407472 is accepted. This one is quite easy to detect, however....

**Comment by carmocca** [comment]
Landing #25110 might be too big of a change, for the specific issue described in the top post. If possible, I would suggest going with using a very negative value instead of -inf when a boolean mask is passed.

**Comment by nikitaved** [comment]
@carmocca , this will still overflow when computing `x - x.max()`.

**Comment by amjames** [comment]
@nikitaved It seems to me this should be okay. The different semantics is permissable as it stems from the behavior of `torch.sparse.softmax` where the namespace `torch.sparse` allows us to have documented departures in semantics under dense<->sparse conversions.

The behavior would still be similar in a predictable, logical way.  That is, for a bsr mask with bool data type the `-inf` would be used automatically in the case of materialized `false` values and unspecified values, and for the float bsr mask the user would need to set materialized values to `-inf`. The assumption of the value of the unspecified elements is a "feature" of the  `sparse.softmax` which has special masking semantics. When the mask is applied weather it be bool or float, the result which eventually makes its way into the softmax has float datatype and the unspecified values are treated the same way. Now with a block sparse mask if the intention is to partially mask some blocks then those specified values must be set appropriately, as you would have to in the case of a dense mask, where all values are specified.

**Comment by pfeatherstone** [comment]
I've got the same bug. It's taken me all day to find this.

**Comment by drisspg** [comment]
I will mark this as high Pri and push a fix that changes the numeric limit used.

**Comment by drisspg** [comment]
Essentially the same Issue: https://github.com/pytorch/pytorch/issues/41508

**Comment by nairbv** [comment]
another small example/repro:
```
>>> t = torch.ones( (5,5), dtype=torch.float).unsqueeze(0)
>>> mask = torch.zeros( (5,5), dtype=torch.float )
>>> mask.fill_( float('-inf'))
tensor([[-inf, -inf, -inf, -inf, -inf],
        [-inf, -inf, -inf, -inf, -inf],
        [-inf, -inf, -inf, -inf, -inf],
        [-inf, -inf, -inf, -inf, -inf],
        [-inf, -inf, -inf, -inf, -inf]])
>>> torch.nn.functional.scaled_dot_product_attention(t, t, t, attn_mask=mask)
tensor([[[nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan]]])
```

In LLMs we have a workaround for this case using:
```
        with torch.no_grad():
            attn = torch.nan_to_num(attn, nan=0.0)
```
but it's unfortunate and a bit confusing since the grad of nan_to_num would cause problems during training too.

I can see semantic arguments about why we might not want to change the behavior in softmax, but I don't think those arguments necessarily translate to this particular function. Can we change the behavior of just the sdpa kernels to return 0's instead of nans?

The function is documented as beta. Maybe we'd also add the `torch.nan_to_num` to the "equivalent to the following" portion of the doc?

**Comment by pfeatherstone** [comment]
I think this is expected behaviour. Look at softmax() of an entirely masked row. You're either going to get inf, Nan or all 1s. The last one is bad because you get past to future leakage. At least the first two "break" your training loop

**Comment by lucidrains** [comment]
@pfeatherstone ah, it isn't my observation, this is just a common early gotcha when it comes to transformer masking

**Comment by nairbv** [comment]
Shouldn't sdpa be able to support a pad token with all-false or (all -inf) attention?

> Look at softmax() of an entirely masked row.

It's the expected behavior of softmax, and this happens to be implemented with a softmax. I'm not seeing why it would be undesirable to change the implementation (of sdpa, not of softmax).

I think this is basically a duplicate but, I agree that we shouldn't be rigid about the behavior of sdpa (or mha) just because of the expected behavior of a function used to implement it: 
https://github.com/pytorch/pytorch/issues/41508#issuecomment-1504840075

**Comment by yacineMTB** [comment]
scaled dot product attention probably wants to apply the mask _after_ the softmax operation when the mask type is boolean

**Comment by vince62s** [comment]
Hi all, any update on this ? shall we use an ugly patch as in transformers in the meantime ?

**Comment by drisspg** [comment]
Since https://github.com/pytorch/pytorch/pull/133882 has landed and will be included in the new 2.5 release I will be closing this isssue no, thanks all for the lively disucssion!
